worker/generalworker.go

```diff
@@ -0,0 +1,409 @@
+			if err != nil {
+				fmt.Println("do something")
+			}
+			client.Transport = newTransport
+
+			*redirectChain = append(
+				*redirectChain,
+				strconv.Itoa(req.Response.StatusCode)+" "+req.URL.String(),
+			)
+			return nil
+		},
+	}
+
+	return client, nil
+}
+
+func MakeRequest(parsedURL *url.URL) (*http.Request, error) {
+	req, err := http.NewRequest("GET", parsedURL.String(), nil)
+	if err != nil {
+		err := fmt.Errorf("request creation error: %w", err)
+		return nil, err
+	}
+	req.Header.Set("User-Agent", "Web Measure/1.0 (https://webresearch.eecs.umich.edu/overview-of-web-measurements/) Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36")
+	req.Header.Set("Cache-Control", "no-cache")
+	return req, nil
+}
+
+// HandleTask is the CORE function of this scanner.
+//
+//	(url sanity) -> check robotstxt cache/DB -> make client
+//	IF NEEDED -> (make robots request -> client.Do<req> -> robots db logging -> make new client ) ->
+//	-> make task request -> client.Do<req> -> resp/err logging
+func (gw *GeneralWorker) HandleTask(task cm.Task, wg *sync.WaitGroup) {
+	var client *http.Client
+	var req *http.Request
+	var resp *http.Response
+	var err error
+
+	defer func() {
+		gw.RespLog(&task, resp)
+		gw.ErrorLog(&task, &err)
+		*gw.AllResultsRef <- task
+		(*wg).Done()
+	}()
+
+	// scheme sanity check
+	parsedURL, _ := url.Parse(task.URL)
+	if parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
+		err = fmt.Errorf("WRONG Scheme! (not http or https)")
+		return
+	}
+
+	// robot.txt check
+	robotsCached := false
+	if !gw.bypassRobots {
+		group := gw.RBConn.Get(parsedURL.Scheme, parsedURL.Hostname())
+		if group != nil {
+			robotsCached = true
+			if !group.Test(parsedURL.Path) {
+				err = fmt.Errorf("path %s not allowd by robots", parsedURL.Path)
+				return
+			}
+		}
+	}
+
+	var redirectChain *[]string
+	if task.Retry == nil {
+		redirectChain = &task.RedirectChain
+	} else {
+		redirectChain = &task.Retry.RedirectChain
+	}
+	*redirectChain = []string{}
+
+	client, err = MakeClient(parsedURL, redirectChain)
+	if err != nil {
+		return
+	}
+
+	if !gw.bypassRobots && !robotsCached {
+		req, _ = MakeRequest(&url.URL{
+			Scheme: parsedURL.Scheme,
+			Host:   parsedURL.Hostname(),
+			Path:   "/robots.txt",
+		})
+		resp, err = client.Do(req)
+
+		var robotsGroup *robotstxt.Group
+		var respBodyStr string
+		var expiration time.Time
+		if err != nil {
+			robotsGroup = &robotstxt.Group{}
+		} else {
+			defer resp.Body.Close()
+			body, _ := io.ReadAll(resp.Body)
+			respBodyStr = string(body)
+			var robotsRaw *robotstxt.RobotsData
+			robotsRaw, err = robotstxt.FromResponse(resp)
+			if err != nil {
+				robotsGroup = &robotstxt.Group{}
+			} else {
+				robotsGroup = robotsRaw.FindGroup("*")
+				expireStr := resp.Header.Get("Expires")
+				if expireStr != "" {
+					expiration, _ = time.Parse(time.RFC1123, resp.Header.Get("Expires"))
+				}
```

Comment:
Probably not super important, cache-control HTTP response headers can also decide the cache expiration date: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expires